File: typescript/packages/postgres/src/artifacts.ts
import {
  type ArtifactStore,
  READ_ONLY,
  type StoreDriver,
  StoreError,
  sha256Hex,
  verified,
} from "@threads/core/store-driver";
import { z } from "zod";
import type { PgDriver } from "./driver";

// Artifacts as rows of the `artifacts` table: content-addressed, re-hashed on every read, and
// shared by every machine on the database. created_at is the putting process's clock, refreshed
// by a re-put (as a re-link refreshes a file's mtime), and gc's grace window is measured
// against it. A put runs in its own transaction, before the event that names it commits.
// ponytail: a bytea is held whole on put and get (scope cut 3); a streaming blob store is the
// follow-up for multi-gigabyte artifacts.

/** gc reads candidates a page at a time, so no scan holds a long serializable transaction. */
const PAGE = 1000;

const Sha = z.array(z.object({ sha256: z.string() }));
const Bytes = z.array(z.object({ bytes: z.instanceof(Uint8Array) }));

export function pgArtifacts(
  driver: PgDriver,
  now: () => number = Date.now,
): ArtifactStore {
  const db = { transaction: driver.detached };
  const put = async (bytes: Uint8Array): Promise<string> => {
    const sha256 = sha256Hex(bytes);
    const stored = Sha.parse(
      await db.transaction((tx) =>
        tx.all(
          `INSERT INTO artifacts (sha256, size, bytes, created_at) VALUES (?, ?, ?, ?)
            ON CONFLICT (sha256) DO UPDATE SET created_at = EXCLUDED.created_at
            RETURNING encode(sha256(bytes), 'hex') AS sha256`,
          [sha256, bytes.length, bytes, now()],
        ),
      ),
    );
    // A row already there holding other bytes: as a file failing its hash on EEXIST.
    if (stored[0]?.sha256 !== sha256)
      throw new StoreError(`artifact ${sha256} is stored with other bytes`);
    return sha256;
  };
  return {
    put,
    get: async (sha256) => {
      const rows = Bytes.parse(
        await db.transaction(
          (tx) =>
            tx.all("SELECT bytes FROM artifacts WHERE sha256 = ?", [sha256]),
          READ_ONLY,
        ),
      );
      const found = rows[0]?.bytes;
      return verified(
        sha256,
        found === undefined ? undefined : new Uint8Array(found),
      );
    },
    sink: () => {
      const chunks: Uint8Array[] = [];
      return {
        write: (chunk) => {
          chunks.push(chunk.slice());
        },
        finish: async () => {
          const bytes = Buffer.concat(chunks);
          return { sha256: await put(bytes), bytes: bytes.length };
        },
        abort: () => {
          chunks.length = 0;
        },
      };
    },
    sweep: async (keep, olderThan) => {
      const removed: string[] = [];
      let after = "";
      for (;;) {
        const page = await candidates(db, olderThan, after);
        const last = page.at(-1);
        if (last === undefined) return removed;
        after = last;
        removed.push(
          ...(await deleted(
            db,
            page.filter((sha) => !keep.has(sha)),
            olderThan,
          )),
        );
      }
    },
  };
}

type Detached = Pick<StoreDriver, "transaction">;

async function candidates(
  db: Detached,
  olderThan: number,
  after: string,
): Promise<readonly string[]> {
  const rows = Sha.parse(
    await db.transaction(
      (tx) =>
        tx.all(
          `SELECT sha256 FROM artifacts WHERE created_at < ? AND sha256 > ?
            ORDER BY sha256 LIMIT ${PAGE}`,
          [olderThan, after],
        ),
      READ_ONLY,
    ),
  );
  return rows.map((r) => r.sha256);
}

/** Each still older than the window: a re-put since the read refreshed it, and it stays. */
async function deleted(
  db: Detached,
  doomed: readonly string[],
  olderThan: number,
): Promise<readonly string[]> {
  if (doomed.length === 0) return [];
  return db.transaction(async (tx) => {
    const gone: string[] = [];
    for (const sha256 of doomed)
      if (
        (await tx.run(
          "DELETE FROM artifacts WHERE sha256 = ? AND created_at < ?",
          [sha256, olderThan],
        )) === 1
      )
        gone.push(sha256);
    return gone;
  });
}

File: python/src/threads/postgres/artifacts.py
"""Artifacts as rows of the `artifacts` table: content-addressed, re-hashed on every read, and
shared by every machine on the database. `created_at` is the putting process's clock, refreshed
by a re-put, and gc's grace window is measured against it."""

from collections.abc import Callable
from functools import partial
from typing import Final

from threads.log import ParseError
from threads.log.digest import sha256_hex
from threads.result import Err, Ok
from threads.store.artifacts import ArtifactSink
from threads.store.conn import Conn, StoreError, run
from threads.store.sql import blob_of, text_of

PAGE: Final = 1000
"""gc reads candidates a page at a time, so no scan holds a long serializable transaction."""


class PgArtifacts:
    def __init__(self, conn: Conn, clock: Callable[[], int]) -> None:
        self._conn = conn
        self._clock = clock

    def put(self, data: bytes) -> str:
        """Stores the bytes, in their own transaction or the caller's. A row already there is
        verified, as a file already there is, and its `created_at` refreshed."""
        sha, now = sha256_hex(data), self._clock()

        def write(conn: Conn) -> str:
            row = conn.execute(
                "INSERT INTO artifacts (sha256, size, bytes, created_at) VALUES (?, ?, ?, ?)"
                " ON CONFLICT (sha256) DO UPDATE SET created_at = EXCLUDED.created_at"
                " RETURNING encode(sha256(bytes), 'hex')",
                (sha, len(data), data, now),
            ).fetchone()
            return "" if row is None else text_of(row[0])

        if run(self._conn, write) != sha:
            raise StoreError(f"artifact {sha} is stored with other bytes")
        return sha

    def get(self, sha256: str) -> Ok[bytes] | Err[ParseError]:
        row = run(
            self._conn,
            lambda c: c.execute(
                "SELECT bytes FROM artifacts WHERE sha256 = ?", (sha256,)
            ).fetchone(),
            read_only=True,
        )
        if row is None:
            return Err(ParseError("artifact_missing", f"no artifact {sha256}"))
        data = blob_of(row[0])
        if sha256_hex(data) != sha256:
            return Err(ParseError("artifact_corrupt", f"artifact {sha256} fails its hash"))
        return Ok(data)

    def sink(self) -> ArtifactSink:
        # ponytail: a bytea is held whole on put (the spec's scope cut 3); a streaming blob store
        # is the follow-up for multi-gigabyte artifacts.
        return _Sink(self)

    def sweep(self, keep: frozenset[str], older_than: int) -> tuple[str, ...]:
        """A read-only pass lists a page of candidates, then a short transaction deletes each
        one still older than the window: a re-put since the read keeps its artifact."""
        removed: list[str] = []
        after = ""
        while page := run(
            self._conn, partial(candidates, older=older_than, after=after), read_only=True
        ):
            after = page[-1]
            doomed = [sha for sha in page if sha not in keep]
            removed += run(self._conn, partial(delete_older, doomed=doomed, older=older_than))
        return tuple(removed)


def candidates(conn: Conn, *, older: int, after: str) -> list[str]:
    rows = conn.execute(
        "SELECT sha256 FROM artifacts WHERE created_at < ? AND sha256 > ? ORDER BY sha256 LIMIT ?",
        (older, after, PAGE),
    ).fetchall()
    return [text_of(sha) for (sha,) in rows]


def delete_older(conn: Conn, *, doomed: list[str], older: int) -> list[str]:
    deleted = "DELETE FROM artifacts WHERE sha256 = ? AND created_at < ?"
    return [sha for sha in doomed if conn.execute(deleted, (sha, older)).rowcount == 1]


class _Sink:
    def __init__(self, store: PgArtifacts) -> None:
        self._store = store
        self._data = bytearray()

    def write(self, chunk: bytes) -> None:
        self._data += chunk

    def commit(self) -> str:
        return self._store.put(bytes(self._data))

    def discard(self) -> None:
        self._data.clear()
